Add "offset" parameter to the Scan query.

gianm opened the pull request:
It works by doing the query as normal and then throwing away the first "offset" number of rows on the broker.

Also includes some logic to skip writing the `limit` field in JSON if it's `Long.MAX_VALUE` (internally this is used to signify "unlimited").

**Edit:** Now also includes changes geared towards stabilizing Scan query result ordering, which is required for offset-based pagination to be useful. See https://github.com/apache/druid/pull/10233#issuecomment-672682361

abhishekrb19 commented:
@gianm, this is a great addition! Do you think it also makes sense to add `OFFSET` to other query types, namely timeseries/groupBy/topN queries? 

I could see timeseries queries returning large result sets, so I think it'd be useful to throw away stuff that is not needed by a client, thereby reducing the # of bytes transferred over the wire. What do you think?

gianm commented:
> @gianm, this is a great addition! Do you think it also makes sense to add `OFFSET` to other query types, namely timeseries/groupBy/topN queries?
> 
> I could see timeseries queries returning large result sets, so I think it'd be useful to throw away stuff that is not needed by a client, thereby reducing the # of bytes transferred over the wire. What do you think?

I was thinking of starting with Scan and GroupBy. Timeseries could make sense too. TopN, I'm not as sure about, because the fact that offset needs to bump up limit means that it will start to increase memory use pretty quickly as you go to the later pages.

gianm commented:
Hmm, I just realized that this method of paginating is a bad idea if you're doing time ordering and lots of rows have the same timestamp. It uses a `java.util.PriorityQueue` internally, which isn't stable with regard to rows with the same timestamp. This is fine if you're only doing one query, but it means that offset-based pagination will not be stable either. More thought and test coverage is needed to make sure we handle this case properly.

gianm commented:
> This pull request **introduces 1 alert** when merging [b3f4f83](https://github.com/apache/druid/commit/b3f4f83e497c78cbab646f058b3bc13028cb75f4) into [34a4113](https://github.com/apache/druid/commit/34a41137522fca463c6af1df5edb2f41297491f5) - [view on LGTM.com](https://lgtm.com/projects/g/apache/druid/rev/pr-1db408ed53e602c8c1de7bd00e4dbe65b7e47bbd)
> 
> **new alerts:**
> 
>     * 1 for Inconsistent equals and hashCode

This is intentional, btw. I added a suppression for the IntelliJ inspection; I'm not sure yet how to suppress the LGTM check.

abhishekrb19 commented:
> I was thinking of starting with Scan and GroupBy. Timeseries could make sense too. TopN, I'm not as sure about, because the fact that offset needs to bump up limit means that it will start to increase memory use pretty quickly as you go to the later pages.

@gianm, yes, that makes sense. Also, I was wondering if in the future we could have a stateful offset implementation where the broker could cache the query/result set from a previous request for a short duration. This could potentially be leveraged in subsequent requests without having to look at the entire result set every time and throw away the rows that are not needed. I think this could get tricky for identical rows, but would be computationally _less_ resource intensive?

gianm commented:
I'd like to get #10235 merged first, then use the StableLimitingSorter from there in this patch.

gianm commented:
> I've converted the patch to a draft.

I've updated the patch to stabilize Scan result order, meaning that it is ready to review. A new test, ScanQueryResultOrderingTest, verifies that for three test segments, the result order is stable at all possible server distributions, ordering, and limit settings.

gianm commented:
> This pull request **introduces 1 alert** when merging [09af824](https://github.com/apache/druid/commit/09af824e1a78eea91978b6a52cce9f8f013193c4) into [e273264](https://github.com/apache/druid/commit/e273264332f5d1e619c50c6ee9727058ccdfb9b2) - [view on LGTM.com](https://lgtm.com/projects/g/apache/druid/rev/pr-3b52efbc134b5ff1588264c5bc52c26b0067a542)
> 
> **new alerts:**
> 
>     * 1 for Inconsistent equals and hashCode

I'm not sure why this warning keeps firing. I tried to add a suppression, but it doesn't seem to work…

Here it is, if anyone's got ideas: https://github.com/apache/druid/pull/10233/files#diff-b67af07d1cac62b329208a8397a0544fR410

suneet-s commented:
> I'm not sure why this warning keeps firing. I tried to add a suppression, but it doesn't seem to work…

It looks correct to me... According to lgtm, the suppressions are only picked up after the change is merged, so I wouldn't worry about the warning - https://lgtm.com/help/lgtm/alert-suppression

gianm commented:
> It looks correct to me... According to lgtm, the suppressions are only picked up after the change is merged, so I wouldn't worry about the warning - https://lgtm.com/help/lgtm/alert-suppression

Ah, I didn't realize that. I was expecting it to respect the suppressions from the particular branch we're on. OK, I'll stop worrying about it then.

abhishekrb19 commented on processing/src/main/java/org/apache/druid/query/scan/ScanQuery.java:
```diff
@@ -202,12 +210,38 @@ public int getBatchSize()
+  }
+
+  /**
+   * Offset for this query; behaves like SQL "LIMIT". Will always be positive. {@link Long#MAX_VALUE} is used in
```
Same comment as https://github.com/apache/druid/pull/10235#discussion_r469416176. :-)

gianm replied:
Yes, you're right, fixed.
